
<!DOCTYPE html
  PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head>
      <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
   <!--
This HTML was auto-generated from MATLAB code.
To make changes, update the MATLAB code and republish this document.
      --><title>ecest580_lab3</title><meta name="generator" content="MATLAB 9.5"><link rel="schema.DC" href="http://purl.org/dc/elements/1.1/"><meta name="DC.date" content="2019-01-31"><meta name="DC.source" content="ecest580_lab3.m"><style type="text/css">
html,body,div,span,applet,object,iframe,h1,h2,h3,h4,h5,h6,p,blockquote,pre,a,abbr,acronym,address,big,cite,code,del,dfn,em,font,img,ins,kbd,q,s,samp,small,strike,strong,sub,sup,tt,var,b,u,i,center,dl,dt,dd,ol,ul,li,fieldset,form,label,legend,table,caption,tbody,tfoot,thead,tr,th,td{margin:0;padding:0;border:0;outline:0;font-size:100%;vertical-align:baseline;background:transparent}body{line-height:1}ol,ul{list-style:none}blockquote,q{quotes:none}blockquote:before,blockquote:after,q:before,q:after{content:'';content:none}:focus{outine:0}ins{text-decoration:none}del{text-decoration:line-through}table{border-collapse:collapse;border-spacing:0}

html { min-height:100%; margin-bottom:1px; }
html body { height:100%; margin:0px; font-family:Arial, Helvetica, sans-serif; font-size:10px; color:#000; line-height:140%; background:#fff none; overflow-y:scroll; }
html body td { vertical-align:top; text-align:left; }

h1 { padding:0px; margin:0px 0px 25px; font-family:Arial, Helvetica, sans-serif; font-size:1.5em; color:#d55000; line-height:100%; font-weight:normal; }
h2 { padding:0px; margin:0px 0px 8px; font-family:Arial, Helvetica, sans-serif; font-size:1.2em; color:#000; font-weight:bold; line-height:140%; border-bottom:1px solid #d6d4d4; display:block; }
h3 { padding:0px; margin:0px 0px 5px; font-family:Arial, Helvetica, sans-serif; font-size:1.1em; color:#000; font-weight:bold; line-height:140%; }

a { color:#005fce; text-decoration:none; }
a:hover { color:#005fce; text-decoration:underline; }
a:visited { color:#004aa0; text-decoration:none; }

p { padding:0px; margin:0px 0px 20px; }
img { padding:0px; margin:0px 0px 20px; border:none; }
p img, pre img, tt img, li img, h1 img, h2 img { margin-bottom:0px; } 

ul { padding:0px; margin:0px 0px 20px 23px; list-style:square; }
ul li { padding:0px; margin:0px 0px 7px 0px; }
ul li ul { padding:5px 0px 0px; margin:0px 0px 7px 23px; }
ul li ol li { list-style:decimal; }
ol { padding:0px; margin:0px 0px 20px 0px; list-style:decimal; }
ol li { padding:0px; margin:0px 0px 7px 23px; list-style-type:decimal; }
ol li ol { padding:5px 0px 0px; margin:0px 0px 7px 0px; }
ol li ol li { list-style-type:lower-alpha; }
ol li ul { padding-top:7px; }
ol li ul li { list-style:square; }

.content { font-size:1.2em; line-height:140%; padding: 20px; }

pre, code { font-size:12px; }
tt { font-size: 1.2em; }
pre { margin:0px 0px 20px; }
pre.codeinput { padding:10px; border:1px solid #d3d3d3; background:#f7f7f7; }
pre.codeoutput { padding:10px 11px; margin:0px 0px 20px; color:#4c4c4c; }
pre.error { color:red; }

@media print { pre.codeinput, pre.codeoutput { word-wrap:break-word; width:100%; } }

span.keyword { color:#0000FF }
span.comment { color:#228B22 }
span.string { color:#A020F0 }
span.untermstring { color:#B20000 }
span.syscmd { color:#B28C00 }

.footer { width:auto; padding:10px 0px; margin:25px 0px 0px; border-top:1px dotted #878787; font-size:0.8em; line-height:140%; font-style:italic; color:#878787; text-align:left; float:none; }
.footer p { margin:0px; }
.footer a { color:#878787; }
.footer a:hover { color:#878787; text-decoration:underline; }
.footer a:visited { color:#878787; }

table th { padding:7px 5px; text-align:left; vertical-align:middle; border: 1px solid #d6d4d4; font-weight:bold; }
table td { padding:7px 5px; text-align:left; vertical-align:top; border:1px solid #d6d4d4; }





  </style></head><body><div class="content"><h2>Contents</h2><div><ul><li><a href="#1">ECES T580 Lab 3 - Tyler Bradley</a></li><li><a href="#2">Lab 3.1.1:</a></li><li><a href="#3">Lab 3.2.1:</a></li><li><a href="#4">Lab 3.2.2:</a></li><li><a href="#5">Lab 3.2.3:</a></li><li><a href="#6">Lab 3.2.3:</a></li><li><a href="#7">Lab 3.3.1:</a></li></ul></div><h2 id="1">ECES T580 Lab 3 - Tyler Bradley</h2><pre class="codeinput">clc;close <span class="string">all</span>; clear;
</pre><h2 id="2">Lab 3.1.1:</h2><p>1. load the sequence in dinoDNA.txt in Matlab.</p><pre class="codeinput">dino_dna = textread(<span class="string">"lab3_files/dinoDNA.txt"</span>, <span class="string">"%q"</span>);

dino_seq = strcat(dino_dna{2:length(dino_dna)});

<span class="keyword">if</span> isfile(<span class="string">"lab3_files/dinoDNA.fasta"</span>)
    delete <span class="string">lab3_files/dinoDNA.fasta</span>
<span class="keyword">end</span>

fastawrite(<span class="string">"lab3_files/dinoDNA.fasta"</span>, <span class="string">"Dino DNA"</span>, dino_seq);

dino_seq = fastaread(<span class="string">"lab3_files/dinoDNA.fasta"</span>);
</pre><h2 id="3">Lab 3.2.1:</h2><p>1. BLAST mystery_sequence.txt online.</p><pre class="codeinput"><span class="comment">% The best match was to "Francisella tularensis subsp. novicida strain</span>
<span class="comment">% AL97-2214, complete genome" with a 68% identity and a E value = 3e-07</span>
</pre><h2 id="4">Lab 3.2.2:</h2><p>1. BLAST dinoDNA.txt in Matlab.</p><pre class="codeinput">[dino_blastn, dino_ROTE] = blastncbi(dino_seq, <span class="string">"blastn"</span>, <span class="string">"database"</span>, <span class="string">"nr"</span>);
dino_bn = getblast(dino_blastn, <span class="string">"ToFile"</span>, <span class="string">"lab3_files/dino_blast.rpt"</span>, <span class="string">"WaitTime"</span>, 1);

<span class="comment">% The best blast match</span>
dino_bn.Hits(1).Definition
</pre><pre class="codeoutput">Blast results are not available yet. Please wait ...

ans =

    'Gallus gallus GATA binding protein 1 (globin transcription factor 1) (GATA1), mRNA &gt;gi|212628|gb|M26209.1|CHKRERYF1 Chicken erythroid-specific transcription factor eryf1 mRNA, complete cds'

</pre><h2 id="5">Lab 3.2.3:</h2><p>Run the code in Ex 3.2.2 and answer the following questions: 1. What was the difference between the blastn and tblastx searches? 2. Why were the shorter lengths chosen? 3. For the full length BLAST, what organisms were closely related? What taxa do these belong to? 4. Did this sequence share homology (similarity) with any known and functionally annotated genes? 5. What was the difference between the tblastx searches performed by the full-length sequence vs. 300 bp sequence vs. 40 bp sequence?</p><pre class="codeinput"><span class="comment">% See answers below blast code</span>

<span class="comment">% run blastn on the mystery sequence to get the answer to 3.</span>
Mys_datan = blastncbi(<span class="string">"lab3_files/mys_seq.fasta"</span>, <span class="string">"blastn"</span>, <span class="string">"database"</span>, <span class="string">"nr"</span>);
mys_blast = getblast(Mys_datan, <span class="string">"WaitTime"</span>, 1);

<span class="comment">% Perform a tblastx and retrieve the report</span>
mys_dna = textread(<span class="string">"lab3_files/mystery_sequence.txt"</span>, <span class="string">"%q"</span>);

mys_seq = strcat(mys_dna{2:length(mys_dna)});

<span class="keyword">if</span> isfile(<span class="string">"lab3_files/mys_seq.fasta"</span>)
    delete <span class="string">lab3_files/mys_seq.fasta</span>
<span class="keyword">end</span>

fastawrite(<span class="string">"lab3_files/mys_seq.fasta"</span>, <span class="string">"Mystery Sequence"</span>, mys_seq);

mys_seq = fastaread(<span class="string">"lab3_files/mys_seq.fasta"</span>);
[Data_tblastx, RTOE_tx] = blastncbi(<span class="string">"lab3_files/mys_seq.fasta"</span>, <span class="string">'tblastx'</span>, <span class="string">'database'</span>, <span class="string">'nr'</span>);
tblastx = getblast(Data_tblastx, <span class="string">"WaitTime"</span>, 1);


<span class="comment">% Perform a tblastx for the first 40 base pairs</span>
Seq_40 = mys_seq.Sequence(1:40);
<span class="keyword">if</span> isfile(<span class="string">"lab3_files/seq_40.fasta"</span>)
    delete <span class="string">lab3_files/seq_40.fasta</span>
<span class="keyword">end</span>
fastawrite(<span class="string">'lab3_files/seq_40.fasta'</span>, <span class="string">'Sequence'</span>, Seq_40);
Seq_40 = fastaread(<span class="string">'lab3_files/seq_40.fasta'</span>);
[Data_tblastx_40, RTOE_40] = blastncbi(<span class="string">"lab3_files/seq_40.fasta"</span>, <span class="string">'tblastx'</span>, <span class="string">'database'</span>, <span class="string">'nr'</span>);
blast_40 = getblast(Data_tblastx_40, <span class="string">"WaitTime"</span>, 1);
<span class="comment">% Perform a tblastx for the first 300 base pairs</span>
Seq_300 = mys_seq.Sequence(1:300);
<span class="keyword">if</span> isfile(<span class="string">"lab3_files/seq_300.fasta"</span>)
    delete <span class="string">lab3_files/seq_300.fasta</span>
<span class="keyword">end</span>
fastawrite(<span class="string">'lab3_files/seq_300.fasta'</span>, <span class="string">'Sequence'</span>, Seq_300);
Seq_300 = fastaread(<span class="string">'lab3_files/seq_300.fasta'</span>);
[Data_tblastx_300, RTOE_300] = blastncbi(<span class="string">"lab3_files/seq_300.fasta"</span>, <span class="string">'tblastx'</span>, <span class="string">'database'</span>, <span class="string">'nr'</span>);
blast_300 = getblast(Data_tblastx_300, <span class="string">"WaitTime"</span>, 1);


<span class="comment">% Answers:</span>
<span class="comment">% 1. The difference between the two blast methods is that blastn compares</span>
<span class="comment">% the sequence to reference gene's nucleotides while tblastx is translating</span>
<span class="comment">% the nucleotide sequence into amino acid and comparing the results to</span>
<span class="comment">% proteins in ncbi's database</span>
<span class="comment">%</span>
<span class="comment">% 2. Shorter lengths were chosen because they were comparing amino acids</span>
<span class="comment">% rather than nucleotide and every amino acid represents a combination of 3</span>
<span class="comment">% nucleotides</span>
<span class="comment">%</span>
<span class="comment">% 3. The top closest match is shown here</span>
mys_blast.Hits(1)

<span class="comment">% 4. The top match to the tblastx blast showing the closest</span>
<span class="comment">% functional genes to the sequence is shown here</span>
tblastx.Hits(1)

<span class="comment">% 5. Looking at the different lengths of the sequence. The 40 bp length did</span>
<span class="comment">% not result in any significant hits in the database for tblastx, which is</span>
<span class="comment">% not surprising given that it is only comparing 13 Aminor acids. The full</span>
<span class="comment">% length sequence had a best hit as shown in number 4 of:</span>
<span class="comment">% 'Niveispirillum cyanobacteriorum strain TH16 chromosome eg_1, complete</span>
<span class="comment">% sequence'.</span>
<span class="comment">% The 300 bp sequence had a best match of</span>
<span class="comment">% 'Indioceanicola profundi strain SCSIO 08040 chromosome, complete genome'.</span>
<span class="comment">% So if different lengths of the sequence are used, it can result in</span>
<span class="comment">% different results from the blast algorithms. The effect of this kind of</span>
<span class="comment">% shortening is likely to be reduced if one looks for regions of the genome</span>
<span class="comment">% that are either highly different between species such as the 16S rRNA gene in</span>
<span class="comment">% prokaryotes or the 18S rRNA gene in eukayorotes</span>
</pre><pre class="codeoutput">Blast results are not available yet. Please wait ...
Blast results are not available yet. Please wait ...
</pre><pre class="codeoutput error">Error using getblast (line 85)
BLAST 56ZESKYT015 is unavailable. Try again later.

Error in ecest580_lab3 (line 61)
tblastx = getblast(Data_tblastx, "WaitTime", 1);
</pre><h2 id="6">Lab 3.2.3:</h2><p>Answer the following questions based on the BLAST results in Lab 3.2.2: 1. What are the top hits of this BLAST search?</p><pre class="codeinput">top_hits = repelem(<span class="string">"a"</span>, 10);
<span class="keyword">for</span> i = 1:10
    top_hits(i) = dino_bn.Hits(i).Definition;
<span class="keyword">end</span>
top_hits

<span class="comment">% 2. What organism do you think Mark used for his dinoDNA sequence.</span>

<span class="comment">% The top hits for the blast results correspond to X.laevis which is a</span>
<span class="comment">% african clawed frog. This match is found in numerous hits in the top ten</span>
<span class="comment">% blast matches</span>

<span class="comment">% Now let?s look for Mark?s hidden message. To do this we need to start a new translated blastx search</span>
<span class="comment">% (blastx). This time use the SwissProt protein database (Matlab parameter: ?database?, ?swissprot?).</span>
<span class="comment">% Answer the questions:</span>

[dino_prot_blast, RTOE_dino_prot] = blastncbi(<span class="string">"lab3_files/dinoDNA.fasta"</span>, <span class="string">'blastx'</span>, <span class="string">'database'</span>, <span class="string">'swissprot'</span>);
dino_tblast = getblast(dino_prot_blast, <span class="string">"WaitTime"</span>, 1);

<span class="comment">% 3. Are the top hits for the blastx search the same as the ones you saw for blastn? Why might this be</span>
<span class="comment">% the case?</span>
<span class="comment">% No they are not the same, the blastx search results correspond to a</span>
<span class="comment">% transcription factor protein that is found in different animals. The top</span>
<span class="comment">% hits for the tblastx correspond to this protein in different animals.</span>
<span class="comment">% This could be different than the nucleotide blastn results because the</span>
<span class="comment">% blastx method makes the assumption that the DNA being used is from a</span>
<span class="comment">% coding region of the genome that would actually be transcribed and</span>
<span class="comment">% translated into a polypeptide chain of amino acids and that may not be</span>
<span class="comment">% true. This assumption may result in inaccurate blast matches</span>

<span class="comment">% 4. What is the hidden message that Mark put in the sequence?</span>
<span class="comment">% This hidden message can be found in the protein sequence of the dino dna</span>
<span class="comment">% and it says "MARK WAS HERE"</span>
regexp(nt2aa(dino_seq.Sequence), <span class="string">"MARK|WAS|HERE"</span>);
regexp(nt2aa(dino_seq.Sequence), <span class="string">"MARK|WAS|HERE"</span>, <span class="string">"match"</span>);
</pre><h2 id="7">Lab 3.3.1:</h2><p>1. Retrieve Peptide Sequences from NCBI (human: AAD01939; fly: AAQ67266).</p><pre class="codeinput">human_pep = getgenpept(<span class="string">'AAD01939'</span>);
fly_pep = getgenpept(<span class="string">'AAQ67266'</span>);
<span class="comment">% 2. Show the Sequence Dot Plot of these two sequences.</span>
seqdotplot(human_pep, fly_pep);
title(<span class="string">"Sequence Dot Plot"</span>);

<span class="comment">% 3. Implement global alignment (Needleman-Wunsch Algorithm) to align these two sequences and</span>
<span class="comment">% show the Scoring space and wining path.</span>
[s_AA,a_AA] = nwalign(human_pep,fly_pep,<span class="string">'Alphabet'</span>,<span class="string">'AA'</span>,<span class="string">'showscore'</span>, <span class="string">'true'</span>);
</pre><p class="footer"><br><a href="https://www.mathworks.com/products/matlab/">Published with MATLAB&reg; R2018b</a><br></p></div><!--
##### SOURCE BEGIN #####
%% ECES T580 Lab 3 - Tyler Bradley
clc;close all; clear;

%% Lab 3.1.1:
% 1. load the sequence in dinoDNA.txt in Matlab.
dino_dna = textread("lab3_files/dinoDNA.txt", "%q");

dino_seq = strcat(dino_dna{2:length(dino_dna)});

if isfile("lab3_files/dinoDNA.fasta")
    delete lab3_files/dinoDNA.fasta
end

fastawrite("lab3_files/dinoDNA.fasta", "Dino DNA", dino_seq);

dino_seq = fastaread("lab3_files/dinoDNA.fasta");

%% Lab 3.2.1:
% 1. BLAST mystery_sequence.txt online.

% The best match was to "Francisella tularensis subsp. novicida strain
% AL97-2214, complete genome" with a 68% identity and a E value = 3e-07


%% Lab 3.2.2:
% 1. BLAST dinoDNA.txt in Matlab.
[dino_blastn, dino_ROTE] = blastncbi(dino_seq, "blastn", "database", "nr");
dino_bn = getblast(dino_blastn, "ToFile", "lab3_files/dino_blast.rpt", "WaitTime", 1);

% The best blast match
dino_bn.Hits(1).Definition

%% Lab 3.2.3:
% Run the code in Ex 3.2.2 and answer the following questions:
% 1. What was the difference between the blastn and tblastx searches?
% 2. Why were the shorter lengths chosen?
% 3. For the full length BLAST, what organisms were closely related? What taxa do these belong to?
% 4. Did this sequence share homology (similarity) with any known and functionally annotated genes?
% 5. What was the difference between the tblastx searches performed by the full-length sequence vs.
% 300 bp sequence vs. 40 bp sequence?

% See answers below blast code 

% run blastn on the mystery sequence to get the answer to 3. 
Mys_datan = blastncbi("lab3_files/mys_seq.fasta", "blastn", "database", "nr");
mys_blast = getblast(Mys_datan, "WaitTime", 1);

% Perform a tblastx and retrieve the report
mys_dna = textread("lab3_files/mystery_sequence.txt", "%q");

mys_seq = strcat(mys_dna{2:length(mys_dna)});

if isfile("lab3_files/mys_seq.fasta")
    delete lab3_files/mys_seq.fasta
end

fastawrite("lab3_files/mys_seq.fasta", "Mystery Sequence", mys_seq);

mys_seq = fastaread("lab3_files/mys_seq.fasta");
[Data_tblastx, RTOE_tx] = blastncbi("lab3_files/mys_seq.fasta", 'tblastx', 'database', 'nr');
tblastx = getblast(Data_tblastx, "WaitTime", 1);


% Perform a tblastx for the first 40 base pairs
Seq_40 = mys_seq.Sequence(1:40);
if isfile("lab3_files/seq_40.fasta")
    delete lab3_files/seq_40.fasta
end
fastawrite('lab3_files/seq_40.fasta', 'Sequence', Seq_40);
Seq_40 = fastaread('lab3_files/seq_40.fasta');
[Data_tblastx_40, RTOE_40] = blastncbi("lab3_files/seq_40.fasta", 'tblastx', 'database', 'nr');
blast_40 = getblast(Data_tblastx_40, "WaitTime", 1);
% Perform a tblastx for the first 300 base pairs
Seq_300 = mys_seq.Sequence(1:300);
if isfile("lab3_files/seq_300.fasta")
    delete lab3_files/seq_300.fasta
end
fastawrite('lab3_files/seq_300.fasta', 'Sequence', Seq_300);
Seq_300 = fastaread('lab3_files/seq_300.fasta');
[Data_tblastx_300, RTOE_300] = blastncbi("lab3_files/seq_300.fasta", 'tblastx', 'database', 'nr');
blast_300 = getblast(Data_tblastx_300, "WaitTime", 1);


% Answers:
% 1. The difference between the two blast methods is that blastn compares
% the sequence to reference gene's nucleotides while tblastx is translating
% the nucleotide sequence into amino acid and comparing the results to
% proteins in ncbi's database
%
% 2. Shorter lengths were chosen because they were comparing amino acids
% rather than nucleotide and every amino acid represents a combination of 3
% nucleotides
%
% 3. The top closest match is shown here
mys_blast.Hits(1)

% 4. The top match to the tblastx blast showing the closest
% functional genes to the sequence is shown here
tblastx.Hits(1)

% 5. Looking at the different lengths of the sequence. The 40 bp length did
% not result in any significant hits in the database for tblastx, which is
% not surprising given that it is only comparing 13 Aminor acids. The full
% length sequence had a best hit as shown in number 4 of:
% 'Niveispirillum cyanobacteriorum strain TH16 chromosome eg_1, complete
% sequence'.
% The 300 bp sequence had a best match of 
% 'Indioceanicola profundi strain SCSIO 08040 chromosome, complete genome'.
% So if different lengths of the sequence are used, it can result in
% different results from the blast algorithms. The effect of this kind of
% shortening is likely to be reduced if one looks for regions of the genome
% that are either highly different between species such as the 16S rRNA gene in
% prokaryotes or the 18S rRNA gene in eukayorotes 

%% Lab 3.2.3:
% Answer the following questions based on the BLAST results in Lab 3.2.2:
% 1. What are the top hits of this BLAST search?
%
top_hits = repelem("a", 10);
for i = 1:10
    top_hits(i) = dino_bn.Hits(i).Definition;
end
top_hits

% 2. What organism do you think Mark used for his dinoDNA sequence.

% The top hits for the blast results correspond to X.laevis which is a
% african clawed frog. This match is found in numerous hits in the top ten
% blast matches

% Now let?s look for Mark?s hidden message. To do this we need to start a new translated blastx search
% (blastx). This time use the SwissProt protein database (Matlab parameter: ?database?, ?swissprot?).
% Answer the questions:

[dino_prot_blast, RTOE_dino_prot] = blastncbi("lab3_files/dinoDNA.fasta", 'blastx', 'database', 'swissprot');
dino_tblast = getblast(dino_prot_blast, "WaitTime", 1);

% 3. Are the top hits for the blastx search the same as the ones you saw for blastn? Why might this be
% the case?
% No they are not the same, the blastx search results correspond to a
% transcription factor protein that is found in different animals. The top
% hits for the tblastx correspond to this protein in different animals.
% This could be different than the nucleotide blastn results because the
% blastx method makes the assumption that the DNA being used is from a
% coding region of the genome that would actually be transcribed and
% translated into a polypeptide chain of amino acids and that may not be
% true. This assumption may result in inaccurate blast matches

% 4. What is the hidden message that Mark put in the sequence?
% This hidden message can be found in the protein sequence of the dino dna
% and it says "MARK WAS HERE"
regexp(nt2aa(dino_seq.Sequence), "MARK|WAS|HERE");
regexp(nt2aa(dino_seq.Sequence), "MARK|WAS|HERE", "match");



%% Lab 3.3.1:
% 1. Retrieve Peptide Sequences from NCBI (human: AAD01939; fly: AAQ67266).
human_pep = getgenpept('AAD01939');
fly_pep = getgenpept('AAQ67266');
% 2. Show the Sequence Dot Plot of these two sequences.
seqdotplot(human_pep, fly_pep);
title("Sequence Dot Plot");

% 3. Implement global alignment (Needleman-Wunsch Algorithm) to align these two sequences and
% show the Scoring space and wining path.
[s_AA,a_AA] = nwalign(human_pep,fly_pep,'Alphabet','AA','showscore', 'true');

##### SOURCE END #####
--></body></html>